### PR TITLE
Test evaluation authorizarion

### PR DESCRIPTION
**Pull request by SepsiLaszlo**

### CODE REVIEW

File: app/controllers/point_details_controller.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class PointDetailsController < ApplicationController
   before_action :require_resort_or_group_leader
-  before_action :require_application_season_for_group_leader
+  # before_action :require_application_season_for_group_leader
```
Comment (balintkiraly) [inline app/controllers/point_details_controller.rb:3]: We should just remove this line instead of the comment. It will be in the git history if we need it later.

**Review by balintkiraly — CHANGES_REQUESTED** [review]

File: app/controllers/point_details_controller.rb (hunk 1 above)
Comment (SepsiLaszlo) [reply app/controllers/point_details_controller.rb]: True, I removed it!

File: spec/requests/entry_requests_spec.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
+require 'rails_helper'
```
Comment (falusi94) [inline spec/requests/entry_requests_spec.rb:1]: Please remove this `require`, it is already included in the rspec. You can add the `# frozen_string _literal: true` instead.

File: spec/requests/entry_requests_spec.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+RSpec.describe "EntryRequests", type: :request do
```
Comment (falusi94) [inline spec/requests/entry_requests_spec.rb:3]:
```suggestion
describe "EntryRequests" do
```

I note it here that we were talking about, it is great that you covered the authorization with tests, but I'd rather keep it as a temporary step that should not be merged into the master. When the other PR is merged all of these authorization rules will be tested twice. I'd recommend to keep here one for the 'happy path' and later on you can add one for a failed authorization, where you can verify the policy is called before the action.

File: spec/requests/entry_requests_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  let(:evaluation) { entry_request.evaluation }
+  let(:group) { evaluation.group }
+  let(:selected_user) { entry_request.user }
+  let(:arguments) do
+    ["/groups/#{group.id}/evaluations/#{evaluation.id}/entryrequests/update",
+     params: {
+         user_id:    selected_user.id,
+         entry_type: EntryRequest::AB
+     }]
+  end
```
Comment (falusi94) [inline spec/requests/entry_requests_spec.rb:14]: DRY-ing up the test is great, but in this case I think it's too much. I constantly have to scroll to-and-fro to get what happens.

File: spec/requests/entry_requests_spec.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  before(:each) do
+    login_as(current_user)
+  end
+
+  shared_context "application season" do
+    before(:each) do
+      SystemAttribute.update_season(SystemAttribute::APPLICATION_SEASON)
+    end
+  end
```
Comment (falusi94) [inline spec/requests/entry_requests_spec.rb:24]: This shared context seems to be too much for me.

File: spec/requests/entry_requests_spec.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+  describe "#update" do
+    include_context "application season"
+
+
+    context "when the user is not authorized" do
+      let(:current_user) { create(:user) }
+
+      it "it's forbidden" do
```
Comment (falusi94) [inline spec/requests/entry_requests_spec.rb:33]:
```suggestion
      it 'returns forbidden' do
```

Please try to avoid duplication of the 'it' in the description. I see it a several places.

File: spec/requests/entry_requests_spec.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        post *arguments
+
+        expect(response).to have_http_status(:forbidden)
+      end
+    end
+
+    context "when the current_user is the group leader" do
+      let(:current_user) { group.leader.user }
+
+      it "it updates" do
```
Comment (falusi94) [inline spec/requests/entry_requests_spec.rb:43]: What do you think about more descriptive cases, like
```suggestion
      it 'updates the entry request' do
```
I think the whole suit's value can be increased like this.

File: spec/requests/point_details_spec.rb [hunk 8]
```diff
@@ -0,0 +1,301 @@
+require 'rails_helper'
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:1]: Same here about the unnecessary require.

File: spec/requests/point_details_spec.rb [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+
+RSpec.describe "PointDetails", type: :request do
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:3]:
```suggestion
describe "PointDetails" do
```

File: spec/requests/point_details_spec.rb [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    let(:arguments) do
+      ["/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+       params: {
+           user_id:       point_detail_user.id,
+           principle_id:  principle.id,
+           evaluation_id: evaluation.id,
+           point:         point
+       },
+       headers: { "ACCEPT" => "application/js" }]
+    end
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:21]: I have the above mentioned feelings about the DRY-ness.

File: test/factories/groups.rb [hunk 11]
```diff
@@ -4,20 +4,25 @@
     type { :group }
     sequence(:id, 400)
 
+    after(:create) do |group|
+      group.memberships << create(:membership, :leader, group: group)
+    end
+
```
Comment (falusi94) [inline test/factories/groups.rb:10]: Is there any reason to add leader to the `basic_group`? I think this should be as lightweight as possible and the leader should remain in the `group`.

File: test/factories/groups.rb [hunk 12]
```diff
@@ -4,20 +4,25 @@
...
-  factory :group, parent: :basic_group do
-    after(:create) do |group|
-      group.memberships << create(:membership, :leader, group: group)
+
+    factory :group_with_parent, parent: :basic_group do
+      parent { create(:basic_group) }
     end
   end
 
+  factory :group, parent: :group_with_parent
```
Comment (falusi94) [inline test/factories/groups.rb:24]: Also, if you move the leader to the `basic_group`, this trait does not make sense. I'd recommend to revert this change and keep the inheritance.

File: test/factories/groups.rb (hunk 11 above)
Comment (SepsiLaszlo) [reply test/factories/groups.rb:10]: i added leader to the `basic_group` because I needed the `group.parent.leader.user` to work in order to test resort leader access.

File: spec/requests/entry_requests_spec.rb (hunk 4 above)
Comment (SepsiLaszlo) [reply spec/requests/entry_requests_spec.rb]: Yes, in this case it could be unnecessary.

File: spec/requests/point_details_spec.rb (hunk 10 above)
Comment (SepsiLaszlo) [reply spec/requests/point_details_spec.rb]: In this case, I think it's not that confusing. The idea behind it that, you don't have to know how the arguments are put together, they are just set.

File: test/factories/groups.rb (hunk 11 above)
Comment (falusi94) [reply test/factories/groups.rb:10]: In this case I'd recommend to change the related factory/initialization to use the `group` instead of the `basic_group`. My problem here that you cannot create a lightweight group w/o leader if it's necessary.

File: spec/requests/point_details_spec.rb [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+      end
+    end
+
+    context "when the current user is the group leader" do
+      let(:current_user) { group.leader.user }
+
+      context "when application season" do
+        include_context "application season"
+
+        context "when the point detail already excists" do
```
Comment (balintkiraly) [inline spec/requests/point_details_spec.rb]:
```suggestion
        context "when the point detail already exists" do
```

**Review by balintkiraly — APPROVED** [review]

File: spec/requests/evalutations_controller_spec.rb [hunk 14]
```diff
@@ -40,4 +71,65 @@
...
+
+        expect(response).to redirect_to(root_url)
+      end
+    end
+
+    context "when the current user is the group leader and application season" do
+      include_context "current user is the group leader"
+      include_context "application season"
+
+      it "it redirects" do
```
Comment (falusi94) [inline spec/requests/evalutations_controller_spec.rb:128]:
```suggestion
      it "redirects" do
```

File: spec/requests/point_details_spec.rb [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+                 params:  {
+                     user_id:       point_detail_user.id,
+                     principle_id:  principle.id,
+                     evaluation_id: evaluation.id,
+                     point:         point
+                 },
+                 headers: { "ACCEPT" => "application/js" }
+
+            expect(PointDetail.last.point_request.user).to eql(point_detail_user)
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:88]:
```suggestion
            expect(PointDetail.last.point_request.user).to eql(point_request.user)
```

File: spec/requests/point_details_spec.rb [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+      it "it returns forbidden" do
+        post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+             params:  {
+                 user_id:       point_detail_user.id,
+                 principle_id:  principle.id,
+                 evaluation_id: evaluation.id,
+                 point:         point
+             },
+             headers: { "ACCEPT" => "application/js" }
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:26]:
```suggestion
        post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update", params: params, headers: headers
```

File: spec/requests/point_details_spec.rb [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        context "when the point detail already exists" do
+          it "it updates the point detail" do
+            post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+                 params:  {
+                     user_id:       point_detail_user.id,
+                     principle_id:  principle.id,
+                     evaluation_id: evaluation.id,
+                     point:         point
+                 },
+                 headers: { "ACCEPT" => "application/js" }
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:47]:
```suggestion
        post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update", params: params, headers: headers
```

File: spec/requests/point_details_spec.rb [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+          it "it doesn't change the PointDetail count" do
+            expect { post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+                          params:  {
+                              user_id:       point_detail_user.id,
+                              principle_id:  principle.id,
+                              evaluation_id: evaluation.id,
+                              point:         point
+                          },
+                          headers: { "ACCEPT" => "application/js" } }.to_not change { PointDetail.count }
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:60]:
```suggestion
            expect do
              post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update", params: params, headers: headers
            end.to_not change { PointDetail.count }
```

File: spec/requests/point_details_spec.rb [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+          it "it has the correct attributes" do
+            post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+                 params:  {
+                     user_id:       point_detail_user.id,
+                     principle_id:  principle.id,
+                     evaluation_id: evaluation.id,
+                     point:         point
+                 },
+                 headers: { "ACCEPT" => "application/js" }
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:71]:
```suggestion
        post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update", params: params, headers: headers
```

File: spec/requests/point_details_spec.rb [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+          it "belong to the correnct user" do
+            post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update",
+                 params:  {
+                     user_id:       point_detail_user.id,
+                     principle_id:  principle.id,
+                     evaluation_id: evaluation.id,
+                     point:         point
+                 },
+                 headers: { "ACCEPT" => "application/js" }
```
Comment (falusi94) [inline spec/requests/point_details_spec.rb:86]:
```suggestion
          post "/groups/#{group.id}/evaluations/#{evaluation.id}/pointdetails/update", params: params, headers: headers
```

**Review by falusi94 — COMMENTED** [review]
Great job, I like how you refactored the factory. Please remove the duplications of 'it's (`it "it does`).

What do you think about a bit of refactoring in the `spec/requests/point_details_spec.rb` too? If I didn't miss something the params building is the same for all cases and I think in and case it's cleaner to extract it to the beginning in one piece instead of passing variables that are not always clear where they come from.

**Review by SepsiLaszlo — COMMENTED** [review]
I think I made all the requested changes.

**Review by falusi94 — APPROVED** [review]
🚀

**Review by OmTheTurtle — APPROVED** [review]
